src/test/groovy/com/rabbitmq/http/client/ReactorNettyClientSpec.groovy

```diff
@@ -1754,6 +1758,138 @@ class ReactorNettyClientSpec extends Specification {
+
+        then: "list of upstreams that contains the new upstream is returned"
+        verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+        cleanup:
+        client.deleteUpstream(vhost, upstreamName).block()
+    }
+
+    def "GET /api/parameters/federation-upstream declare and get at non-root vhost"() {
+        given: "an upstream"
+        final vhost = "foo"
+        final upstreamName = "upstream2"
+        client.createVhost(vhost).block()
+        declareUpstream(client, vhost, upstreamName)
+
+        when: "client requests the upstreams"
+        final upstreams = awaitEventPropagation { client.getUpstreams(vhost) }
+
+        then: "list of upstreams that contains the new upstream is returned"
+        verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+        cleanup:
+        client.deleteUpstream(vhost, upstreamName).block()
+        client.deleteVhost(vhost).block()
+    }
+
+    def "PUT /api/parameters/federation-upstream with null upstream uri"() {
+        given: "an Upstream without upstream uri"
+        UpstreamDetails upstreamDetails = new UpstreamDetails()
+
+        when: "client tries to declare an Upstream"
+        client.declareUpstream("/", "upstream3", upstreamDetails).block()
+
+        then: "an illegal argument exception is thrown"
+        thrown(IllegalArgumentException)
+    }
+
+    def "DELETE /api/parameters/federation-upstream/{vhost}/{name}"() {
+        given: "upstream upstream4 in vhost /"
+        final vhost = "/"
+        final upstreamName = "upstream4"
+        declareUpstream(client, vhost, upstreamName)
+
+        def upstreams = awaitEventPropagation { client.getUpstreams() }
+        verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+        when: "client deletes upstream upstream4 in vhost /"
+        client.deleteUpstream(vhost, upstreamName).block()
+
+        and: "upstream list in / is reloaded"
+        upstreams = client.getUpstreams()
+
+        then: "upstream4 no longer exists"
+        !upstreams.filter{ it.name.equals(upstreamName) }.hasElements().block()
+    }
+
+    def "GET /api/parameters/federation-upstream-set declare and get"() {
+        given: "an upstream set with two upstreams"
+        final vhost = "/"
+        final upstreamSetName = "upstream-set-1"
+        final upstreamA = "A"
+        final upstreamB = "B"
+        final policyName = "ppp"
+        declareUpstream(client, vhost, upstreamA);
+        declareUpstream(client, vhost, upstreamB);
+        final d1 = new UpstreamSetDetails()
+        d1.setUpstream(upstreamA)
+        d1.setExchange("amq.direct")
+        final d2 = new UpstreamSetDetails()
+        d2.setUpstream(upstreamB)
+        d2.setExchange("amq.fanout")
+        final detailsSet = new ArrayList()
+        detailsSet.add(d1)
+        detailsSet.add(d2)
+        client.declareUpstreamSet(vhost, upstreamSetName, detailsSet).block()
+        PolicyInfo p = new PolicyInfo()
```

Comment:
Any reasons there's a policy implied in this test and not in the synchronous client test?

Reply:
Added policy to the synchronous client test as well.
The point behind adding a policy to the test is to have a demo of creating a federation.